### PR TITLE
[DEV APPROVED] 8720 stamp duty calculator be feature

### PR DESCRIPTION
**Pull request by willhall88**
https://moneyadviceservice.tpondemand.com/entity/8720

This feature adds the back end functionality for new rates and standards that have been released with the latest government budget. 

First time buyer rates and thresholds have now been changed.

As all as adding this functionality, there has been quite a large refactor of the code to hopefully make it more maintainable in the future.

### CODE REVIEW

**Comment by willhall88** [comment]
There is one failing test. deliberately failing as the test needs to be fixed. This is however outside of the scope of this PR and will be addressed with a subsequent ticket

**Comment by Tr4pSt3R** [comment]
@willhall88 rather than leaving it as a failing spec can we make it a pending test with a reasonable explanation on what needs doing please? Along with sufficient information for anyone who will pick it up at a later date

**Comment by willhall88** [comment]
@Tr4pSt3R I was doing that just as you commented! 🙈  - it is the next piece of work that will be picked up so I will ensure the knowledge isn't lost

**Comment by Tr4pSt3R** [comment]
Oh great! Thanks @willhall88

**Comment by willhall88** [comment]
Apologies for the large size of this PR. lots of refactoring and had to extend a lot of the tests to get them all fixed and working. 
All cucumber tests need to run javascript and non-javascript, these can definitely be refactored soon. but at least this implementation runs all the tests, which they weren't doing before

File: app/models/mortgage_calculator/stamp_duty.rb [hunk 1]
```diff
@@ -79,5 +70,38 @@ def percentage_tax
...
       result.nan? ? 0 : result
     end
+
+    def second_home?
+      buyer_type == 'isSecondHome'
+    end
+
+    private
+
+    def add_tax_band(band, rate)
```
Comment (margOnline) [inline app/models/mortgage_calculator/stamp_duty.rb]: can we have a different name for this method, perhaps `tax_for_band`, `band_tax`

File: app/models/mortgage_calculator/stamp_duty.rb [hunk 2]
```diff
@@ -1,70 +1,61 @@
...
-      end
+      bands = first_time_rate? ? FIRST_TIME_BUYER_BANDS : STANDARD_BANDS
 
-      if self.second_home && price >= SECOND_HOME_THRESHOLD
-        total_tax += price * (SECOND_HOME_RATE / 100)
+      bands.each do |band|
+        rate = band[tax_rate]
+        total_tax += add_tax_band(band, rate)
+        break if price_in_band?(band)
       end
```
Comment (margOnline) [inline app/models/mortgage_calculator/stamp_duty.rb]: this method is so much more readable 👍

File: app/models/mortgage_calculator/stamp_duty.rb [hunk 3]
```diff
@@ -1,70 +1,61 @@
...
 
     currency_inputs :price
-
     validates :price, presence: true, numericality: true
 
     def initialize(options = {})
-      self.price = options.fetch(:price){ 0 }
-      self.second_home = options.key?(:second_home) && options[:second_home] == "true"
+      self.price = options.fetch(:price) { 0 }
+      self.buyer_type = options[:buyer_type] || 'isNextHome'
```
Comment (Tr4pSt3R) [inline app/models/mortgage_calculator/stamp_duty.rb]: @willhall88 freeze this `'isNextHome'` value perhaps?

File: spec/models/stamp_duty_spec.rb [hunk 4]
```diff
@@ -6,37 +6,37 @@
       expect(subject.price).to be_zero
     end
 
-    it 'sets second_home to false' do
-      expect(subject.second_home).to be_falsy
+    it 'sets buyer_type to standard by default' do
+      expect(subject.buyer_type).to eq'isNextHome'
```
Comment (Tr4pSt3R) [inline spec/models/stamp_duty_spec.rb]: `expect(subject.buyer_type).to eq'isNextHome'` --> `expect(subject.buyer_type).to eq('isNextHome')`

File: spec/models/stamp_duty_spec.rb [hunk 5]
```diff
@@ -6,37 +6,37 @@
...
   end
 
-  describe '#second_home' do
-    subject { described_class.new(second_home: checkbox_value) }
+  describe '#second_home?' do
+    subject { described_class.new(buyer_type: buyer_type) }
 
     context 'when "false" is given' do
-      let(:checkbox_value) { "false" }
+      let(:buyer_type) { "isFTB" }
```
Comment (Tr4pSt3R) [inline spec/models/stamp_duty_spec.rb]: `let(:buyer_type) { "isFTB" }` --> `let(:buyer_type) { 'isFTB' }`

**Comment by willhall88** [comment]
:) thanks folks

**Pull request merged**